### PR TITLE
Reactor: ensure consistent vector lengths in C++ layer

### PR DESCRIPTION
**Pull request by ischoegl**
**Changes proposed in this pull request**

- Ensure consistent vector sizes, i.e. `Reactor::m_advancelimits.size()` is `Reactor::m_nv`
- Prevents hard to track segfaults when `m_advancelimits.size()` < `m_nv` in custom additions
- Make functions related to advance limits non-virtual

### CODE REVIEW

**Review by speth — COMMENTED** [review]
I see the issue with the keeping the length of `m_advancelimits` consistent across different derived classes (including user-created ones). I wanted to propose an alternative solution, which is to simply resize it to the correct size in the few locations where it is used (all of which are within the `Reactor` base class). See bd58afa72 for the implementation of this option.

I'm :+1: on making the related methods non-virtual.

**Comment by ischoegl** [comment]
@speth ... resizing at a later time is a *very* good suggestion! Not having to resize in overloaded functions is great. Along those lines, I am suggesting to not resize `m_advancelimits` at all unless it is actually being used. (which allows for simple checks that reduce overhead).

File: src/zeroD/Reactor.cpp [hunk 1]
```diff
@@ -445,20 +444,25 @@ void Reactor::setAdvanceLimits(const double *limits)
         throw CanteraError("getState",
                            "Error: reactor is empty.");
     }
-    for (size_t j = 0; j < m_nv; j++) {
-        m_advancelimits[j] = limits[j];
+    m_advancelimits.resize(m_nv, -1.0);
+    std::copy(limits, limits + m_nv, m_advancelimits.begin());
```
Comment (speth) [inline src/zeroD/Reactor.cpp]: In one line:
```c++
m_advancelimits.assign(limits, limits + m_nv);
```

File: src/zeroD/Reactor.cpp [hunk 2]
```diff
@@ -445,20 +444,25 @@ void Reactor::setAdvanceLimits(const double *limits)
...
+
+    // resize to zero length if no limits are set
+    if (std::none_of(m_advancelimits.begin(), m_advancelimits.end(),
+                     [](double l){return l>0;})) {
```
Comment (speth) [inline src/zeroD/Reactor.cpp]: Can we have a variable name besides `l`? Looks way too much like `1` in many fonts. Applies to both instances of this.

File: src/zeroD/Reactor.cpp [hunk 3]
```diff
@@ -445,20 +444,25 @@ void Reactor::setAdvanceLimits(const double *limits)
...
 bool Reactor::getAdvanceLimits(double *limits)
 {
-    bool has_limit = false;
-
-    for (size_t j = 0; j < m_nv; j++) {
-        limits[j] = m_advancelimits[j];
-        has_limit |= (limits[j] > 0.);
+    // m_advancelimit has zero length if no limits are set
+    bool has_limit = m_advancelimits.size();
+    if (has_limit) {
```
Comment (speth) [inline src/zeroD/Reactor.cpp:459]: Using `hasAdvanceLimits()` here would be self-documenting.

**Review by speth — CHANGES_REQUESTED** [review]
This looks good to me. I only have a couple of minor suggestions. In addition to these, can you squash the third commit into the first, since it reverts a number of the changes made in the first commit?

**Comment by ischoegl** [comment]
Done (and squashed).

**Review by speth — APPROVED** [review]

**Pull request merged**